### PR TITLE
Always use 0 for offsets of lib / extern union members

### LINKED ISSUE #9744: Compiler bug: invalid Int32 when calling offsetof(lib union)
```crystal
{"" => [] of Int8}
{"" => [] of Int16}
{"" => [] of Int32}
{"" => [] of Int64}
{"" => [] of UInt8}
{"" => [] of UInt16}
{"" => [] of UInt32}

lib LibT
  union U
    x : Int32
    y : UInt8
  end
end

def instance_var_names(type : T.class) forall T
  {{ T.instance_vars.map(&.stringify) }}
end

p instance_var_names(LibT::U)
p offsetof(LibT::U, @y)
```

This snippet leads to the following compiler error on x86_64-unknown-linux-gnu 52c750ece (LLVM 8.0.0):

```
Invalid Int32: 93825018395136 (ArgumentError)
  from ../../../../crystal/src/string.cr:428:5 in 'to_i32'
  from ../../../../crystal/src/compiler/crystal/codegen/codegen.cr:448:23 in 'accept'
  from ../../../../crystal/src/compiler/crystal/codegen/codegen.cr:2193:7 in 'prepare_call_args_non_external'
  from ../../../../crystal/src/compiler/crystal/codegen/call.cr:57:7 in 'visit'
  from ../../../../crystal/src/compiler/crystal/syntax/visitor.cr:27:12 in 'accept'
  from ../../../../crystal/src/compiler/crystal/codegen/codegen.cr:628:9 in 'accept'
  from ../../../../crystal/src/compiler/crystal/codegen/codegen.cr:2158:7 in 'codegen'
  from ../../../../crystal/src/compiler/crystal/compiler.cr:172:16 in 'compile'
  from ../../../../crystal/src/compiler/crystal/command.cr:210:5 in 'run_command'
  from ../../../../crystal/src/compiler/crystal/command.cr:62:5 in 'run'
  from ../../../../crystal/src/crystal/main.cr:105:5 in 'main'
  from src/env/__libc_start_main.c:94:2 in 'libc_start_main_stage2'
Error: you've found a bug in the Crystal compiler. Please open an issue, including source code that will allow us to reproduce the bug: https://github.com/crystal-lang/crystal/issues
```

On 0.35.1 the snippet compiles, but it also produces [very incorrect results](https://carc.in/#/r/9p12), as `offsetof(LibT::U, @y)` should be 0 because it is a C union.

The invalid integer is different every time the snippet is run; it seems to be some kind of corrupted pointer.  What I believe is that the compiler tries to fix the value of `offsetof` before all types are known and finalized.

### PR DESCRIPTION
**Pull request by HertzDevil**
Fixes #9744. See https://github.com/crystal-lang/crystal/issues/13205#issuecomment-1503847741. Removes the workaround in #7335.

`pointerof` doesn't have the same issue. Also it seems interpreter specs were broken for a while and fixed here, but I am not 100% sure.

@will @biqqles @zhangkaizhao If possible, please check whether this PR affects #13205, #12589, and https://github.com/crystal-lang/crystal/issues/11934#issuecomment-1079851964 respectively

### CODE REVIEW

**Comment by HertzDevil** [comment]
It seems the requirement for `current_debug_file` was dropped in LLVM 13: https://reviews.llvm.org/rGad60802a7187aa39b0374536be3fa176fe3d6256

**Review by straight-shoota — APPROVED** [review]

File: src/llvm/target_data.cr [hunk 1]
```diff
@@ -23,6 +23,8 @@ struct LLVM::TargetData
   end
 
   def offset_of_element(struct_type, element)
+    # element_count = LibLLVM.count_struct_element_types(struct_type)
+    # raise "Invalid element idx!" unless element < element_count
```
[the remark this answers is not in the document]
Comment (HertzDevil) [reply src/llvm/target_data.cr:27]: This is intentionally left there, like the rest of the debug stuffs in e.g. `LLVM::Builder`

**Comment by will** [comment]
This does look like it fixes it. I think? I wasn't able to easily get ahold of llvm 10.0.0 (the version the darwin-universal has) to try, and it seems like llvm 10.0.1 also fixes it. So it's a little unsatisfying to not be able to build with 10.0.0 to fully verify. 


```
bash-3.2$ ~/Downloads/crystal-1.7.3-1/bin/crystal version
Crystal 1.7.3 [d61a01e18] (2023-03-07)

LLVM: 10.0.0
Default target: aarch64-apple-darwin
bash-3.2$ ~/Downloads/crystal-1.7.3-1/bin/crystal build --debug test.cr
Assertion failed: (Idx < NumElements && "Invalid element idx!"), function getElementOffset, file /var/cache/omnibus/src/llvm/llvm-10.0.0.src/include/llvm/IR/DataLayout.h, line 608.
/Users/will/Downloads/crystal-1.7.3-1/bin/crystal: line 102: 51671 Abort trap: 6           "$INSTALL_DIR/embedded/bin/crystal" "$@"
```

```
bash-3.2$ nix run .#crystal_173_llvm10 -- version
Crystal 1.7.3 [d61a01] (1970-01-01)

LLVM: 10.0.1
Default target: aarch64-apple-darwin
bash-3.2$ nix run .#crystal_173_llvm10 -- build --debug test.cr
```

```
bash-3.2$ nix run .#crystal_patch_llvm10 -- version
Crystal 1.8.0-dev [da38e4] (1970-01-01)

LLVM: 10.0.1
Default target: aarch64-apple-darwin
bash-3.2$ nix run .#crystal_patch_llvm10 -- build --debug test.cr
```

```
bash-3.2$ nix run .#crystal_patch_llvm15 -- version
Crystal 1.8.0-dev [da38e4] (1970-01-01)

LLVM: 15.0.7
Default target: aarch64-apple-darwin
bash-3.2$ nix run .#crystal_patch_llvm15 -- build --debug test.cr
bash-3.2$ nix shell .#crystal_patch_llvm10
```

test.cr is from your comment on my bug report
```crystal
lib Foo
  union Bar
    x : Int32
    y : Int16
    z : Int8
  end
end

offsetof(Foo::Bar, @x) # => 0
offsetof(Foo::Bar, @y) # => 110689
offsetof(Foo::Bar, @z)
```

but it behaves the same as the crunchydata/bridge-cli project with respect to 10.0.1 also seeming to fix the issue, and only failing with the pre-compiled release on github.

flake I used to try all the llvm versions and patches

```nix
{
  inputs = {
    flake-utils.url = "github:numtide/flake-utils";

    nixpkgs-crystal = {
      url = "github:will/nixpkgs-crystal";
      inputs.nixpkgs.follows = "nixpkgs";
    };
  };
  description = "crystal tests";

  outputs = { self, nixpkgs, nixpkgs-crystal, flake-utils }:
    let
      systems = (builtins.attrNames nixpkgs-crystal.outputs.packages);
    in
    flake-utils.lib.eachSystem systems (system:
      let
        pkgs = nixpkgs.legacyPackages.${system};

        base = nixpkgs-crystal.packages.${system}.crystal;
        release_src = pkgs.fetchFromGitHub {
          owner = "crystal-lang";
          repo = "crystal";
          rev = "d61a01e185baf84046efce50832a97f097ba0a2f";
          hash = "sha256-ULhLGHRIZbsKhaMvNhc+W74BwNgfEjHcMnVNApWY+EE=";
        };
        patch_src = pkgs.fetchFromGitHub {
          owner = "HertzDevil";
          repo = "crystal";
          rev = "da38e42a66df7bbc2e6e5a6265f4c1f9cf3f2980";
          hash = "sha256-q4nr8qMMRDxJyJ6eURdcIsRi5dsj6ZQ18V4eXMjrr0U=";
        };

      in
      {
        packages = {
          crystal_173_llvm10 = base.override {
            llvm = pkgs.llvm_10;
            src = release_src;
          };

          crystal_patch_llvm10 = base.override {
            llvm = pkgs.llvm_10;
            src = patch_src;
          };

          crystal_patch_llvm15 = base.override {
            llvm = pkgs.llvm_15;
            src = patch_src;
          };
        };
      }
    );
}
```

**Comment by HertzDevil** [comment]
In that case 10.0.1 might have had assertions disabled

**Comment by biqqles** [comment]
Thanks for the ping, this looks very promising! Unfortunately I can't test properly for Solus as since `1.7.0` compiling std or compiler specs results in

```
make -j8 std_spec CRYSTAL_CONFIG_TARGET=x86_64-linux-gnu
Using /usr/bin/llvm-config [version= 14.0.6]
./bin/crystal build -D strict_multi_assign -D preview_overload_order --release --link-flags="-Wl,--copy-dt-needed-entries -Wl,-O1 -Wl,-z,relro -Wl,-z,now -Wl,-z,max-page-size=0x1000 -Wl,-Bsymbolic-functions -Wl,--sort-common" -Dwithout_interpreter  --exclude-warnings spec/std --exclude-warnings spec/compiler --exclude-warnings spec/primitives -o .build/std_spec spec/std_spec.cr
Using compiled compiler at .build/crystal
Showing last frame. Use --error-trace for full trace.

In src/exception/call_stack/libunwind.cr:1:1

 1 | require "c/dlfcn"
     ^
Error: can't find file 'c/dlfcn'

If you're trying to require a shard:
- Did you remember to run `shards install`?
- Did you make sure you're running the compiler in the same directory as your shard.yml?
make: *** [Makefile:176: .build/std_spec] Error 1
```

Based on my recollections of bringing up the package, it could be that `CRYSTAL_CONFIG_TARGET` is being ignored? The compiler itself builds fine.

**Comment by HertzDevil** [comment]
Does `./bin/crystal env CRYSTAL_PATH` include the `lib` directory in this repository?

**Comment by biqqles** [comment]
Awesome, thank you for saving me some time, adding the `lib:` prefix to `CRYSTAL_PATH` indeed resolved it.

Compiling now...

**Comment by biqqles** [comment]
@HertzDevil or @straight-shoota -- do I need to open an issue for the new assertion failure?

**Comment by straight-shoota** [comment]
This interpreter issue is tracked in #13317

**Pull request merged**